nav2_pure_pursuit/src/pure_pursuit_controller.cpp

```diff
@@ -0,0 +1,414 @@
+  applyKinematicConstraints(
+    linear_vel, angular_vel,
+    fabs(lookahead_dist - sqrt(carrot_dist2)), lookahead_dist, dt.seconds());
+
+  // Collision checking
+  if (isCollisionImminent(pose, carrot_pose)) {
+    RCLCPP_ERROR(logger_, "Collision imminent!");
+    throw std::runtime_error("PurePursuitController detected collision ahead!");
+  }
+
+  // populate and return message
+  geometry_msgs::msg::TwistStamped cmd_vel;
+  cmd_vel.header.frame_id = pose.header.frame_id;
+  cmd_vel.header.stamp = clock_->now();
+  cmd_vel.twist.linear.x = linear_vel;
+  cmd_vel.twist.angular.z = angular_vel;
+  last_cmd_ = cmd_vel;
+  return cmd_vel;
+}
+
+geometry_msgs::msg::PoseStamped PurePursuitController::getLookAheadMarker(
+  const double & lookahead_dist,
+  const nav_msgs::msg::Path & transformed_plan)
+{
+  // Find the first pose which is at a distance greater than the lookahead distance
+  auto goal_pose_it = std::find_if(
+    transformed_plan.poses.begin(), transformed_plan.poses.end(), [&](const auto & ps) {
+      return hypot(ps.pose.position.x, ps.pose.position.y) >= lookahead_dist;
+    });
+
+  // If the no pose is not far enough, take the last pose
+  if (goal_pose_it == transformed_plan.poses.end()) {
+    goal_pose_it = std::prev(transformed_plan.poses.end());
+  }
+
+  return *goal_pose_it;
+}
+
+bool PurePursuitController::isCollisionImminent(
+  const geometry_msgs::msg::PoseStamped & robot_pose,
+  const geometry_msgs::msg::PoseStamped & carrot_pose)
+{
+  // This may be a bit unusual, but the robot_pose is in odom frame
+  // and the carrot_pose is in robot base frame. We need to collision
+  // check in odom frame, so all values will be relative to robot base pose.
+  // But we can still use the carrot pose to find quantities and convert to odom.
+
+  const double carrot_dist = hypot(carrot_pose.pose.position.x, carrot_pose.pose.position.y);
+  const unsigned int num_pts = static_cast<unsigned int>(ceil(carrot_dist / costmap_->getResolution()));
+  const double pt_distance = carrot_dist / num_pts;
+
+  geometry_msgs::msg::PoseStamped carrot_in_odom;
+  try {
+    tf_->transform(carrot_pose, carrot_in_odom,
+      costmap_ros_->getGlobalFrameID(), transform_tolerance_);
+    return true;
+  } catch (tf2::TransformException & ex) {
+    RCLCPP_ERROR(
+      logger_,
+      "Exception attempting to get look ahead point in odom frame: %s", ex.what());
+  }
+
+  geometry_msgs::msg::Vector3 unit_vector;
+  unit_vector.x = carrot_in_odom.pose.position.x / carrot_dist;
+  unit_vector.y = carrot_in_odom.pose.position.y / carrot_dist;
+
+  double curr_dist;
+  geometry_msgs::msg::Vector3 cur_pt;
+
+  for (unsigned int i = 0; i != num_pts; i++) {
+    curr_dist = i * pt_distance;
+    cur_pt.x = robot_pose.pose.position.x + (curr_dist * unit_vector.x);
+    cur_pt.y = robot_pose.pose.position.y + (curr_dist * unit_vector.y);
+    if (inCollision(cur_pt.x, cur_pt.y)) {
```

Comment:
Maybe add footprint checking, at a later stage? Possibly reuse DWB's code for that?

Reply:
I had thought about it, but this collision checking is already pretty approximate. We're collision checking the straight line between the robot and the carrot. In reality, the robot will not drive that way. Since we don't know the exact path the robot will take in reality, we can't know the orientation of the robot base to be able to apply the footprint. 

We could apply a forward simulation to the velocity command instead and ignore the carrot which would allow us to have estimated orientations instead, but its unclear to me how long in time or in space to apply that forward simulation to be complete. I'm not sure if the termination of the velocity command would eventually hit the carrot point in all situations (I think it might) to know when to terminate the forward simulation

Reply:
@shrijitsingh99 you brought up a good point and I rethought collision detection, we're not using it based on the velocities, thanks for the note here, its definitely a better way to go.